microk8s-resources/actions/metrics-server.yaml

```diff
@@ -208,3 +208,29 @@ subjects:
 - kind: ServiceAccount
   name: metrics-server
   namespace: kube-system
+---
+kind: ClusterRole
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: system:aggregated-metrics-reader
+  labels:
+    rbac.authorization.k8s.io/aggregate-to-view: "true"
+    rbac.authorization.k8s.io/aggregate-to-edit: "true"
+    rbac.authorization.k8s.io/aggregate-to-admin: "true"
+rules:
+- apiGroups: ["metrics.k8s.io"]
+  resources: ["pods", "nodes"]
+  verbs: ["get", "list", "watch"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: microk8s-admin
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: admin
+subjects:
+- apiGroup: rbac.authorization.k8s.io
+  kind: User
+  name: 127.0.0.1
```

Comment:
I suspect the 127.0.0.1 has to do with how we create the CA (use 127.0.0.1 for CN), right?

Reply:
That's right

Reply:
Do you think we should be issuing a proxy certificate sign it by our CA and pass that to `--proxy-client-cert-file`?

Reply:
Generating another cert for proxy instead of using `server.crt` won't harm (and generally recommended) however it doesn't have much to do with 127.0.0.1 above. This `127.0.0.1` comes from the cert which is used for kubectl (which is using `ca.crt` for authentication).

Reply:
```
$ kubectl config view --raw -o json | jq -r '.clusters[].cluster."certificate-authority-data"' | base64 -d | openssl x509 -noout -subject
subject=CN = 127.0.0.1
```

Reply:
I actually need to confirm my own understanding.

Reply:
Removed my own wrong comments so as not to confuse. Will check again and come back.

Reply:
> Do you think we should be issuing a proxy certificate sign it by our CA and pass that to --proxy-client-cert-file?

Yes, I think we should.

Reply:
How about this?
```
diff --git a/microk8s-resources/actions/common/utils.sh b/microk8s-resources/actions/common/utils.sh
index 6b390a9..46b52bc 100644
--- a/microk8s-resources/actions/common/utils.sh
+++ b/microk8s-resources/actions/common/utils.sh
@@ -142,8 +142,8 @@ get_ips() {
 }
 
 
-produce_server_cert() {
-    # Produce the server certificates based on the rendered csr.conf.rendered.
+produce_certs() {
+    # Produce certificates based on the rendered csr.conf.rendered.
     # The file csr.conf.rendered is compared with csr.conf to determine if a regeneration of server certs must be done.
     #
     # Returns 
@@ -158,8 +158,12 @@ produce_server_cert() {
 
     if ! "${SNAP}/usr/bin/cmp" -s "${SNAP_DATA}/certs/csr.conf.rendered" "${SNAP_DATA}/certs/csr.conf"; then
       cp ${SNAP_DATA}/certs/csr.conf.rendered ${SNAP_DATA}/certs/csr.conf
+      # Generate the server certificate
       openssl req -new -key ${SNAP_DATA}/certs/server.key -out ${SNAP_DATA}/certs/server.csr -config ${SNAP_DATA}/certs/csr.conf
       openssl x509 -req -in ${SNAP_DATA}/certs/server.csr -CA ${SNAP_DATA}/certs/ca.crt -CAkey ${SNAP_DATA}/certs/ca.key -CAcreateserial -out ${SNAP_DATA}/certs/server.crt -days 100000 -extensions v3_ext -extfile ${SNAP_DATA}/certs/csr.conf
+      # Generate the proxy client certificate
+      openssl req -new -key ${SNAP_DATA}/certs/front-proxy-client.key -out ${SNAP_DATA}/certs/front-proxy-client.csr -config ${SNAP_DATA}/certs/csr.conf -subj "/CN=front-proxy-client"
+      openssl x509 -req -in ${SNAP_DATA}/certs/front-proxy-client.csr -CA ${SNAP_DATA}/certs/ca.crt -CAkey ${SNAP_DATA}/certs/ca.key -CAcreateserial -out ${SNAP_DATA}/certs/front-proxy-client.crt -days 100000 -extensions v3_ext -extfile ${SNAP_DATA}/certs/csr.conf
       echo "1"
     else
       echo "0"
diff --git a/microk8s-resources/wrappers/apiservice-kicker b/microk8s-resources/wrappers/apiservice-kicker
index a7d2b22..42e6106 100755
--- a/microk8s-resources/wrappers/apiservice-kicker
+++ b/microk8s-resources/wrappers/apiservice-kicker
@@ -23,7 +23,7 @@ do
        ip route | grep default &> /dev/null &&
        systemctl is-active --quiet snap.microk8s.daemon-apiserver.service
     then
-        csr_modified="$(produce_server_cert)"
+        csr_modified="$(produce_certs)"
         if [[ "$csr_modified" -eq "1" ]];
         then
             echo "CSR change detected. Reconfiguring the kube-apiserver"
diff --git a/snap/hooks/configure b/snap/hooks/configure
index c49a33b..ea049a7 100755
--- a/snap/hooks/configure
+++ b/snap/hooks/configure
@@ -99,7 +99,7 @@ fi
 # Make sure the server certificate includes the IP we are using
 if ! [ -f ${SNAP_DATA}/certs/csr.conf ]
 then
-    produce_server_cert
+    produce_certs
     rm -rf .srl
     systemctl restart snap.${SNAP_NAME}.daemon-apiserver.service
     systemctl restart snap.${SNAP_NAME}.daemon-proxy.service
diff --git a/snap/hooks/install b/snap/hooks/install
index b71ecef..9b9bff6 100755
--- a/snap/hooks/install
+++ b/snap/hooks/install
@@ -17,7 +17,8 @@ openssl genrsa -out ${SNAP_DATA}/certs/serviceaccount.key 2048
 openssl genrsa -out ${SNAP_DATA}/certs/ca.key 2048
 openssl req -x509 -new -nodes -key ${SNAP_DATA}/certs/ca.key -subj "/CN=127.0.0.1" -days 10000 -out ${SNAP_DATA}/certs/ca.crt
 openssl genrsa -out ${SNAP_DATA}/certs/server.key 2048
-produce_server_cert
+openssl genrsa -out ${SNAP_DATA}/certs/front-proxy-client.key 2048
+produce_certs
 rm -rf .srl
 
 # Create the basic tokens
```

Reply:
Yes! That's what I had in mind.